pkg/kube/client/ssa/caching_unstructured_extractor.go

```diff
@@ -0,0 +1,237 @@
+		return nil, err
+	}
+
+	e.cache.mu.Lock()
+	defer e.cache.mu.Unlock()
+	// invalidate stale entries in cache with the fresh discovery data
+	for gvCached, cacheEntry := range e.cache.store {
+		path := gvRelativeAPIPath(gvCached)
+		if discoGV, ok := oapiPathsToGV[path]; !ok || discoGV.ETag() != cacheEntry.etag {
+			delete(e.cache.store, gvCached)
+		}
+	}
+
+	gvPath := gvRelativeAPIPath(gv)
+	oapiGV, ok := oapiPathsToGV[gvPath]
+	if !ok {
+		return nil, fmt.Errorf("cannot find GroupVersion %q in discovery", gvPath)
+	}
+
+	// check the cache after invalidating stale data
+	parserTuple, ok := e.cache.store[gv]
+	// cache hit
+	if ok && parserTuple.etag == oapiGV.ETag() && oapiGV.ETag() != "" {
+		return parserTuple.parser, nil
+	}
+	// generate new parser on cache miss or etag mismatch
+	// defensively cover the case where discovery does not return any ETag
+	// for GV, which normally should not happen
+	freshParser, err := newParserFromOpenAPIGroupVersion(ctx, oapiGV)
+	if err != nil {
+		return nil, err
+	}
+	// cache parser only if non-empty etag
+	if oapiGV.ETag() != "" {
+		e.cache.store[gv] = &GVKParserCacheEntry{
+			parser: freshParser,
+			etag:   oapiGV.ETag(),
+		}
+	}
+	return freshParser, nil
+}
+
+// gvRelativeAPIPath constructs the OpenAPI path for the given GVK
+func gvRelativeAPIPath(gv schema.GroupVersion) string {
+	if gv.Group == "" {
+		return "api/" + gv.Version
+	}
+	return "apis/" + gv.String()
+}
+
+func newParserFromOpenAPIGroupVersion(ctx context.Context, oapiGV OpenAPIGroupVersion) (*GvkParser, error) {
+	// note: although proto schema is more performant, we are
+	// using the JSON schema here, as there is an issue with
+	// proto.NewOpenAPIV3Data during makeUnions() at
+	// https://github.com/kubernetes/kube-openapi/blob/f7e401e7b4c2199f15e2cf9e37a2faa2209f286a/pkg/schemaconv/smd.go#L128
+	s, err := oapiGV.Schema(ctx, "application/json")
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get OpenAPI schema")
+	}
+	var oapi spec3.OpenAPI
+	if err := json.Unmarshal(s, &oapi); err != nil {
+		return nil, errors.Wrap(err, "cannot unmarshal OpenAPI schema")
+	}
+
+	var refErrors []string
+	// validate that every reference in each schema in the OpenAPI document
+	// is in the document, i.e. OpenAPI document is self-contained
+	// with no unresolvable or external reference.
+	// errors are expected to be accumulated into refErrors,
+	// by the RefCallback function,
+	// as the schema walker has no means of stopping early
+	walker := schemamutation.Walker{
+		SchemaCallback: schemamutation.SchemaCallBackNoop,
+		// note: this should not mutate any ref, only validate
+		RefCallback: validateRefSelfContainedFn(&refErrors, oapi.Components.Schemas),
+	}
+	specs := map[string]*spec.Schema{}
+	for k, v := range oapi.Components.Schemas {
+		walker.WalkSchema(v)
+		specs[k] = v
+	}
+	if len(refErrors) > 0 {
+		return nil, errors.Errorf("cannot validate references in OpenAPI schemas: %s", strings.Join(refErrors, ",\n"))
```

Comment:
why not use kube's `NewAggregate` with `[]error`. Why this manual string acrobatic?

Reply:
done https://github.com/crossplane-contrib/provider-kubernetes/pull/246/commits/81c421cac3d398c3688895af32e0c6fcde1a288e